### ISSUE TITLE
Multiple executions overwrite eachother

### ISSUE THREAD

**Issue by netbrain** [temyers/cucumber-jvm-parallel-plugin#24]
Given the configuration:
```
            <plugin>
                <groupId>com.github.temyers</groupId>
                <artifactId>cucumber-jvm-parallel-plugin</artifactId>
                <version>1.2.1</version>
                <executions>
                    <execution>
                        <id>generateAndroidRunners</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>generateRunners</goal>
                        </goals>
                        <configuration>
                        <configuration>
                            ...
                            <tags>"@android"</tags>
                            <namingScheme>simple</namingScheme>
                        </configuration>
                        </configuration>
                    </execution>
                    <execution>
                        <id>generateIOSRunners</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>generateRunners</goal>
                        </goals>
                        <configuration>
                            ...
                            <tags>"@ios"</tags>
                            <namingScheme>simple</namingScheme>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

If i have a single feature in the project this configuration would only create Parallel01IT.java, where i would expect this to create two files, one with the IOS configuration one with the Android configuration.

However the workaround for now is to have the naming scheme on one of these executions set to feature-title.

I guess something like #19 would be a solution for this, or a naming scheme which is a bit more intelligent.

**Comment by saikrishna321** [comment]
@netbrain why cant we have both android and ios tag together in the Parallel01T.java ?

**Comment by netbrain** [comment]
The first execution has a different set of glue packages than the last one.

This is done in order to have a different configuration for each context, for instance i bootstrap android tests with a AndroidDriver, while ios with a IOSDriver and add capabilities the corresponding driver capabilities. So in order to automatically parallelize them, i require some way of creating test runners with different configuration.

The ability to set a prefix would be sufficient.

**Comment by saikrishna321** [comment]
@netbrain i get it clear on the Driver caps .... 

with the first point, i'm still unclear-- you can still set glue packages common separated 
```
    <!-- comma separated list of package names to scan for glue code -->
         <glue>foo, bar</glue>
```

i run cucumber  tests in parallel, please take a look at this (https://github.com/saikrishna321/AppiumTestDistribution)

Sample Project (https://github.com/saikrishna321/PageObjectPatternAppium)

I have made some slight changes as i wanted the Parallel01T.java to be outside the classpath and i trigger the test in parallel using ExecutorService.. This framework works for both android/ios on real devices in parallel. Hope this helps you.

**Comment by temyers** [comment]
However, in order to execute the tests, you would need a separate execution configuration.

Generally, I'd expect this to be done through the use of Maven Profiles, for example:
```
<profiles>
<profile>
<id>run-android-tests</id>
...
<plugin>
                <groupId>com.github.temyers</groupId>
                <artifactId>cucumber-jvm-parallel-plugin</artifactId>
                <version>1.2.1</version>
                <executions>
                    <execution>
                        <id>generateAndroidRunners</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>generateRunners</goal>
                        </goals>
                        <configuration>
                        <configuration>
                            ...
                            <tags>"@android"</tags>
                            <namingScheme>simple</namingScheme>
                        </configuration>
                        </configuration>
                    </execution>
            </plugin>
</profile>
<profile>
<id>run-ios-tests</id>
<plugin>
                <groupId>com.github.temyers</groupId>
                <artifactId>cucumber-jvm-parallel-plugin</artifactId>
                <version>1.2.1</version>
                <executions>
                    <execution>
                        <id>generateIOSRunners</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>generateRunners</goal>
                        </goals>
                        <configuration>
                            ...
                            <tags>"@ios"</tags>
                            <namingScheme>simple</namingScheme>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
</profile>
```

And each set of tests executed by a different build.

**Comment by temyers** [comment]
Are you wanting to run ALL the tests within a SINGLE build execution?

#19 could give a solution to that.

**Comment by netbrain** [comment]
@saikrishna321 Yes, i am using several packages in the glue statement.

@temyers: Yes i want to run everything in a single build execution, but i could always create profiles as you said and make them run by default. However this bloats the pom.xml file and I really want to keep things as simple as possible.

Now things are working just fine, as long as i have the first execution naming-scheme set to simple and the other set to feature-title. However I would still expect it possible to run several executions of the plugin without the second run overwriting the changes of the last.

**Comment by temyers** [comment]
@saikrishna321: My understanding (and correct me if I'm wrong @netbrain), is that the glue packages need to be different, due to the use of "Inversion of Control".  Each target (Android, iOS) has a separate package implementing the required behaviour.  Therefore the *same* tests are executed against different targets by varying the glue code packages.  Android and iOS tests can be executed in parallel.

**Comment by netbrain** [comment]
Thats correct.

**Comment by saikrishna321** [comment]
@netbrain i get that... would it b possible for you to point to some example of same.. would like to take a look on how its been implemented

**Comment by netbrain** [comment]
Heres a small example (some code is removed for abbreviation purposes):
```

//org.example.common.view.LoginView
public interface LoginView {
    void login();
}

//org.example.ios.view.IOSLoginView
public class IOSLoginView extends TestContext implements LoginView {
    @Override
    public void login() {
        workWithWebViewCtx();
        WebElement email = wd.findElement(By.xpath("//*[@data-tam='inputEmail']"));
        email.sendKeys(AuthConfig.getUsername());

        WebElement pwd = wd.findElement(By.xpath("//*[@data-tam='inputPassword']"));
        pwd.sendKeys(AuthConfig.getPassword());

        WebElement submit = wd.findElement(By.xpath("//*[@data-tam='submitFormButton']"));
        submit.click();

        workWithNativeCtx();
    }

}

//org.example.steps.LoginSteps
public class LoginSteps extends TestContext {

    LoginView loginView = getView(LoginView.class);
        
    @When("^I have logged into my account$")
    public void iHaveLoggedIntoMyAccount() throws Throwable {
        loginView.login();
    }

}

```
The TestContext class is a class with the webdriver as a static variable and a couple of other usefull functionality like, isIOS() or isAndroid(), as well as getView() which fetches the correct implementation of a view depending on if the test is currently runnning on a android or ios device. (i.e either AndroidLoginView or IOSLoginView)

If I were using a dependency injection framework i would have, instead of extending the TestContext class, injected it as a dependency. However, for now it's fine.

**Comment by temyers** [comment]
ok, this is a good use case for #19.  

Each Maven execution is separate so I see the current behaviour (overwriting changes when using the same naming scheme) consistent with how it works with other Maven plugins.

**Comment by netbrain** [comment]
Question, if running this as maven profiles instead. It would still suffer the issue of the test runners being overwritten by eachother? In the instance where I would run a command like the following `mvn test -P ios -P android` or if I made the two profiles run by default.

**Comment by temyers** [comment]
> Question, if running this as maven profiles instead. It would still suffer the issue of the test runners being overwritten by eachother?

Without separate execution then yes.
The plugin is called in a separate process/class loader, which means that things like static fields (used for the simple 1 up counter) are reset between runs.

However,:
mvn test -Pios
mvn test -Pandroid

Wouldn't interfere _as much_.  This would need a caveat that depending on *exactly* what is generated and how tests are executed it's possible that you still might get unexpected results.  In the above scenario, I'd expect that there are a number of other parameters that would reduce the liklihood of overlap (e.g. the tags that are used for the execution).

**Comment by temyers** [comment]
But the plugin should help reduce the chance of conflict.  

And I think naming convention would probably be the simplest solution at this time.
Having to cater for multiple executions would add unnecessary complication IMO.

**Comment by temyers** [comment]
@netbrain Apologies, I was wrong with my earlier comments regarding the Maven classloader.  Having reviewed the code again, the counter I used when generating files was not static (I thought it was).  

I've created a test case which demonstrates the issue and have a fix I'll be deploying shortly as part of issue #19 so you can create multiple executions with simple naming convention that shouldn't overlap.

**Comment by netbrain** [comment]
Good that something came out of this!, however I've myself fallen back to using maven profiles and only running test on a single device. As it seems that reports generated from the tests was just a complete mess as it had no way of distinguishing which features was run on android and which were on ios.

**Comment by temyers** [comment]
Is that down to the parallel plugin or the underlying reporting itself?  That's one reason why I've always separated such test suites, since otherwise you can't ascertain which test was run.  

Perhaps there's an opportunity for another plugin?  To automatically add tags to a suite of tests at runtime?

**Comment by netbrain** [comment]
Im using the allure framework to generate reports, and they group their test results into the feature name. But when running android and ios test in the same execution context, then it would basically aggregate the results of each feature even if their running in a "android" or "ios" context. The allure reporter doesn't know how to distinguish the two, and for good reason. 

Another plugin which could split the features into a android and ios context would probably fix this issue. The plugin would have to duplicate the existing features and prepend a [ANDROID] or [IOS] to the feature name. This way when running the entire test suite, the report should contain test results which are correctly separated. Either that or there needs to be some sort of Allure listener which, during runtime can manipulate the test results by adding context information.

However, the workaround now is to separate android and ios tests into their own execution context.

**Comment by temyers** [comment]
Fixed by pr #33
